### PR TITLE
Add nova, glance and keystone to rjil-cicd

### PR DESCRIPTION
**Pull request by rickysarraf**
This patch adds nova, glance and keystone to the rjil-cicd build framework. It also adapts all packages that had a service in them. It also includes the pre/post housekeeping tasks for each service.

The changes have been verified to build under Ubuntu Trusty chroot.

### CODE REVIEW

**Comment by sorenh** [comment]
What kind of testing have you done on this?

nova-compute-*.* aren't used for anything, afaict?

**Comment by rickysarraf** [comment]
On Wednesday 25 March 2015 04:33 PM, Soren Hansen wrote:
>
> What kind of testing have you done on this?
>
> nova-compute-/./ aren't used for anything, afaict?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JioCloud/autobuild/pull/9#issuecomment-85983098>.
>

It builds fine in my setup, under pbuilder.

The current packaging only builds a single nova-compute deb, and that
deb contains just the upstrat config file.
Right now, there is not individual nova-compute-* package.

rrs@learner:/var/tmp/Debian-Build/Result$ dpkg -I
nova-compute_2014.2.137_all.deb
 new debian package, version 2.0.
 size 5824 bytes: control archive=824 bytes.
      28 bytes,     1 lines      conffiles           
     303 bytes,    11 lines      control             
      74 bytes,     1 lines      md5sums             
     542 bytes,    18 lines   *  postinst             #!/bin/sh
     343 bytes,    11 lines   *  preinst              #!/bin/sh
     219 bytes,     7 lines   *  prerm                #!/bin/sh
 Package: nova-compute
 Source: rjil-cicd
 Version: 1:2014.2.137
 Architecture: all
 Maintainer: Ritesh Raj Sarraf <Ritesh.Sarraf@ril.com>
 Installed-Size: 38
 Depends: rjil-cicd (= 1:2014.2.137)
 Section: admin
 Priority: optional
 Description: nova-compute package
  Automatically generated nova-compute package
19:56 ♒♒♒   ☺   
rrs@learner:/var/tmp/Debian-Build/Result$ dpkg --contents
nova-compute_2014.2.137_all.deb
drwxr-xr-x root/root         0 2015-03-24 16:14 ./
drwxr-xr-x root/root         0 2015-03-24 16:13 ./usr/
drwxr-xr-x root/root         0 2015-03-24 16:13 ./usr/share/
drwxr-xr-x root/root         0 2015-03-24 16:13 ./usr/share/doc/
drwxr-xr-x root/root         0 2015-03-24 16:14
./usr/share/doc/nova-compute/
-rw-r--r-- root/root      4060 2015-03-24 16:10
./usr/share/doc/nova-compute/changelog.gz
drwxr-xr-x root/root         0 2015-03-24 16:14 ./etc/
drwxr-xr-x root/root         0 2015-03-24 16:14 ./etc/init/
-rw-r--r-- root/root       679 2015-03-24 16:06 ./etc/init/nova-compute.conf
19:56 ♒♒♒   ☺   


-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
On Wednesday 25 March 2015 07:58 PM, Ritesh Raj Sarraf wrote:
> The current packaging only builds a single nova-compute deb, and that
> deb contains just the upstrat config file.
> Right now, there is not individual nova-compute-* package.

I just checked the repo. There are pre/post installation scripts for
individual nova-compute-* package, but were ineffective because I didn't
create per-hypervisor pacakges.


-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by sorenh** [comment]
Please don't leave files in the repo that serve no purpose. They only make things more confusing.

**Comment by sorenh** [comment]
Please remove any references to nova-baremetal.

**Comment by sorenh** [comment]
There are some control files. Are they overrideng defaults or why are they included here?

**Comment by rickysarraf** [comment]
On Wednesday 25 March 2015 11:57 PM, Soren Hansen wrote:
> Please don't leave files in the repo that serve no purpose. They only
> make things more confusing. 

I've cleaned this up now.

-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
On Thursday 26 March 2015 12:01 AM, Soren Hansen wrote:
>
> Please remove any references to nova-baremetal.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JioCloud/autobuild/pull/9#issuecomment-86161739>.
>

Done. Removed all references to nova-baremetal.

-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
On Thursday 26 March 2015 12:03 AM, Soren Hansen wrote:
>
> There are some control files. Are they overrideng defaults or why are
> they included here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JioCloud/autobuild/pull/9#issuecomment-86162396>.
>

Oh!!. These were leftover from my previous test build. :-(
I've cleaned them up.

I've also added  a manual clean-up in the clean target to make sure any
.control files doesn't get left over.

-- 
Ritesh Raj Sarraf
RESEARCHUT - http://www.researchut.com
"Necessity is the mother of invention."

**Comment by rickysarraf** [comment]
Shall we get this merged ?

**Comment by sorenh** [comment]
I've not had a chance to take this for a spin, I'm afraid. I'm also on holiday, so I probably won't have time to review it until early next week.

In addition to build testing, can you please work with Dan or Harish to see if we can somehow get this tested functionally?

**Comment by rickysarraf** [comment]
@bodepd @hkumarmk 
Can you please help with testing this change ?

This merge request will introduce 3 new server side OpenStack components into the rjil-cicd deb package.

The first prime test is to ensure that this change does not break the existing setup and workflow.
And the second test will be to verify that the rjil-cicd deb package is pulled in, when we try to install the server packages.

The server packages that this merge introduces are:

glance-api
glance-common
glance-registry
keystone
nova-api-ec2
nova-api-metadata
nova-api-os-compute
nova-api
nova-cells
nova-cert
nova-compute
nova-conductor
nova-consoleauth
nova-console
nova-network
nova-novncproxy
nova-objectstore
nova-scheduler
nova-spiceproxy
nova-xvpncproxy


This is not the final list. If we beleive we need to remove/add some of the packages, we can do that. All these packages will Depend on rjil-cicd deb package. Hence, any installation request on any of these packages will automatically pull in the rjil-cicd packages.

**Comment by rickysarraf** [comment]
Please test this

File: debian/glance-api.prerm [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!/bin/sh
+
+set -e
+
+case $1 in
+	remove|purge)
+		if [ -x /etc/init.d/glance-api ]; then
+			invoke-rc.d glance-api stop || true
```
Comment (sorenh) [inline debian/glance-api.prerm]: Won't debhelper add this on its own?

File: debian/glance-registry.postinst [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+if [ "$1" = "configure" ]
+then
+	if grep -qE "^(sql_)?connection.*sqlite.*" /etc/glance/glance-registry.conf
+	then
+		su -s /bin/sh -c 'glance-manage db_sync' glance
+	fi
+fi
+
+#DEBHELPER#
+
```
Comment (sorenh) [inline debian/glance-registry.postinst]: I don't think we need this?

File: debian/glance-registry.prerm [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+
+set -e
+
+case $1 in
+	remove|purge)
+		if [ -x /etc/init.d/glance-registry ]; then
+			invoke-rc.d glance-registry stop || true
```
Comment (sorenh) [inline debian/glance-registry.prerm]: Again, I'm pretty sure debhelper already does this?

File: debian/keystone.postinst [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		exit 0
+		;;
+esac
+
+# dh_installdeb will replace this with shell code automatically
+# generated by other debhelper scripts.
+
+#DEBHELPER#
+
+exit 0
```
Comment (sorenh) [inline debian/keystone.postinst]: We have an external db, so I don't think we need any of this.

File: debian/nova-common.postinst [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    fi
+
+    if [ "$(id -gn nova)"  = "nogroup" ]; then
+      usermod -g nova nova
+    fi
+
+    if [ -z "$2" ]; then
+      # New install - blanket permissions
+      chown -R nova:nova /var/lib/nova/
+    elif dpkg --compare-versions "$2" lt "2011.3-0ubuntu4"; then
```
Comment (sorenh) [inline debian/nova-common.postinst]: I'm pretty sure we don't support upgrades from versions this old.

File: debian/nova-common.postinst [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      # make sure that LXC rootfs mount points are excleuded
+      # during upgrades from previous versions
+      find /var/lib/nova/ -name 'rootfs' -prune -o \
+        -group root -a -user nova -exec chown nova:nova {} \;
+      find /var/lib/nova/ -name 'rootfs' -prune -o \
+        -group nogroup -a -user nova -exec chown nova:nova {} \;
+    elif dpkg --compare-versions "$2" lt "2012.2~f1~20120503.13935-0ubuntu1"; then
```
Comment (sorenh) [inline debian/nova-common.postinst]: ....or this old.

File: debian/nova-common.postinst [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	  # convert the root_helper to rootwrap_config
+	  sed -e "s,^root_helper=.\+,rootwrap_config=/etc/nova/rootwrap.conf," -i /etc/nova/nova.conf
+    fi	
+
+    if [ "$(id -gn nova)"  = "nogroup" ]; then
+        usermod -g nova nova
+    fi
+
+    if dpkg --compare-versions "$2" lt "2012.1~e3-0ubuntu2"; then
```
Comment (sorenh) [inline debian/nova-common.postinst]: ...or this old.

File: debian/nova-common.postinst [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+      usermod -s "/bin/false" nova
+    fi
+
+    chown -R nova:nova /etc/nova
+    chown -R nova:adm /var/log/nova
+
+    if [ -z "$2" ]; then
+        # New install - blanket permissions
+        chown -R nova:nova /var/lib/nova/
+    elif dpkg --compare-versions "$2" lt "2011.3-0ubuntu4"; then
```
Comment (sorenh) [inline debian/nova-common.postinst]: ...or this old.

**Comment by sorenh** [comment]
I see many references to rootwrap config, but no .install file that installs it.

**Comment by rickysarraf** [comment]
Rebased it to master, to include all changes after merge of tempest into master.

**Comment by rickysarraf** [comment]
All of usr/ and etc/ is picked by debian/rjil-cicd.install

**Comment by rickysarraf** [comment]
@sorenh  SInce, we are setting up repos for nova, keystone, glace, neutron now,  how do we plan on having them packaged ? This PR is for having them included into rjil-cicd....

**Comment by anshprat** [comment]
closing this as it has to be taken up by respective engineering teams. This PR would be used as reference for change though.

**Pull request closed without merging**